### PR TITLE
Thread application error bit through HTTP inbound and outbound

### LINKED ISSUE #489: Forwarding: Application error bit
Right now I don't think a tchannel proxy of an http service will be able to set the application error bit becsu that's not sent over the wire for HTTP.

### PR DESCRIPTION
**Pull request by kriskowal**
This adds support for the application error bit on call responses for both TChannel and HTTP. For TChannel, this bit already transited, but for HTTP, it is necessary to add an HTTP response header, Rpc-Status, which can carry the value "error". I left the header absent in the success case for simplicity, but any value other than error lifts to "not error".

### CODE REVIEW

**Comment by abhinav** [comment]
Just throwing an idea out there: What if we call the header `RPC-Status` with
valid values being `success` and `error`?

**Comment by kriskowal** [comment]
Pushing on this today, since we need to land this before end of week, Dec 23.

**Comment by kriskowal** [comment]
@abhinav ok. if anything, leaves room for a Rpc-Error: $case if we later want to add support for deciding whether to retry based on application error case.

File: transport/http/handler.go [hunk 1]
```diff
@@ -197,6 +197,6 @@ func (rw responseWriter) AddHeaders(h transport.Headers) {
 	applicationHeaders.ToHTTPHeaders(h, rw.w.Header())
 }
 
-func (responseWriter) SetApplicationError() {
-	// Nothing to do.
+func (rw responseWriter) SetApplicationError() {
+	rw.w.Header().Add(ApplicationStatusHeader, ApplicationErrorStatus)
```
Comment (abhinav) [inline transport/http/handler.go]: Wouldn't you rather use `Set`? `Set` ensures there's only a single instance, overwriting existing ones.

File: transport/http/constants.go [hunk 2]
```diff
@@ -56,4 +56,12 @@ const (
...
 	// application-specific routing.
 	RoutingDelegateHeader = "Rpc-Routing-Delegate"
+
+	// ApplicationStatusHeader is the HTTP response header used to express
+	// whether the response body represents an application error.
+	ApplicationStatusHeader = "Rpc-Status"
+
+	// ApplicationErrorStatus is the value for ApplicationStatusHeader for
+	// errant requests.
+	ApplicationErrorStatus = "error"
```
Comment (abhinav) [inline transport/http/constants.go]: Should we set the header to `"success"` or `"ok"` for the happy case? It feels icky that the header is present only in the failure case, only with that value.

**Comment by blampe** [comment]
Please leave room to ambiguity client errors from server errors.

File: transport/http/constants.go (hunk 2 above)
Comment (kriskowal) [inline transport/http/constants.go]: It could be done—would need to set it initially and override in SetApplicationError.

**Comment by kriskowal** [comment]
@blampe This change pertains only to application errors, as can only come from the server, not a proxy or client. @abhinav can speak to the point of how we distinguish client errors from server errors. We may need to talk further about how to distinguish proxy errors from server errors—the convention in HTTP is to use gateway instead of server status codes, but we haven’t written anything special.

File: transport/http/handler.go (hunk 1 above)
Comment (kriskowal) [inline transport/http/handler.go]: ✔

File: transport/http/constants.go (hunk 2 above)
Comment (kriskowal) [inline transport/http/constants.go]: ✔

**Pull request merged**